dashboard/app/models/pd/application/teacher1819_application.rb

```diff
@@ -34,5 +34,229 @@ def set_type_and_year
+    validates_presence_of :course
+    before_validation :set_course_from_program
+    def set_course_from_program
+      self.course = PROGRAMS.key(program)
+    end
+
+    before_create :match_partner, if: -> {regional_partner.nil?}
+    def match_partner
+      self.regional_partner = RegionalPartner.find_by_region(zip_code, state_code)
+    end
+
+    PROGRAMS = {
+      csd: 'Computer Science Discoveries (appropriate for 6th - 10th grade)',
+      csp: 'Computer Science Principles (appropriate for 9th - 12th grade, and can be implemented as an AP or introductory course)',
+    }.freeze
+    PROGRAM_OPTIONS = PROGRAMS.values
+
+    GRADES = [
+      'Pre-K'.freeze,
+      'Kindergarten'.freeze,
+      *(1..12).map {|n| "Grade #{n}".freeze}
+    ].freeze
+
+    SUBJECTS_THIS_YEAR = [
+      'Computer Science',
+      'Computer Literacy',
+      'Math',
+      'Science',
+      'History',
+      'Social Studies',
+      'English/Language Arts',
+      'Music',
+      'Art',
+      'Multimedia',
+      'Foreign Language',
+      OTHER_PLEASE_LIST
+    ]
+
+    COURSE_HOURS_PER_YEAR = [
+      'At least 100 course hours',
+      '50 to 99 course hours',
+      'Less than 50 course hours'
+    ]
+
+    TERMS_PER_YEAR = [
+      '1 quarter',
+      '1 trimester',
+      '1 semester',
+      '2 trimesters',
+      'Full year'
+    ]
+
+    def self.options
+      {
+        country: [
+          'United States',
+          'International'
+        ],
+
+        title: COMMON_OPTIONS[:title],
+        state: COMMON_OPTIONS[:state],
+        gender_identity: COMMON_OPTIONS[:gender_identity],
+        race: COMMON_OPTIONS[:race],
+        principal_title: COMMON_OPTIONS[:title],
+
+        current_role: [
+          'Teacher',
+          'Instructional Coach',
+          'Librarian',
+          'School administrator',
+          'District administrator',
+          OTHER_PLEASE_LIST
+        ],
+
+        grades_at_school: GRADES,
+        grades_teaching: [
+          *GRADES,
+          "I'm not teaching this year (please explain):"
+        ],
+        grades_expect_to_teach: GRADES,
+
+        subjects_teaching: SUBJECTS_THIS_YEAR,
+        subjects_expect_to_teach: SUBJECTS_THIS_YEAR,
+
+        subjects_licensed_to_teach: [
+          'Computer Science',
+          'Career and Technical Education',
+          'Math',
+          'Science',
+          'History',
+          'Social Studies',
+          'English/Language Arts',
+          'Music',
+          'Art',
+          'Multimedia',
+          'Foreign Language',
+          'Business',
+          'Special Education',
+          'Physical Education',
+          'I am not currently licensed',
+          OTHER_PLEASE_LIST
+        ],
+
+        taught_in_past: [
+          'Hour of Code',
+          'CS Fundamentals',
+          'CS in Algebra',
+          'CS in Science',
+          'CS Discoveries',
+          'CS Principles (intro or AP-level)',
+          'AP CS A',
```

Comment:
It hasn't been taught in 8 years but should AP CS AB be here?

Reply:
Nope, we don't need to include.